File: RELEASE_BRANCH_371/chem/KPP/kpp/kpp-2.1/int/gillespie.c
void StochasticRates( double RCT[], double Volume, double SCT[] );   
void Propensity ( int V[], int F[], double SCT[], double A[] );
void MoleculeChange ( int j, int NmlcV[] );
double CellMass(double T); 
void Update_RCONST();

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void Gillespie(int Nevents, double Volume, double* T, int NmlcV[], int NmlcF[]) 
/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
{
    
      int i, m=0, event;
      double r1, r2;
      double A[NREACT], SCT[NREACT], x;

     /* Compute the stochastic reaction rates */
      Update_RCONST();
      StochasticRates( RCONST, Volume, SCT );   
   
      for (event = 1; event <= Nevents; event++) {

          /* Uniformly distributed ranfor (m numbers */
          r1 = (double)rand()/(double)RAND_MAX;
          r2 = (double)rand()/(double)RAND_MAX;

	  /* Avoid log of zero */
	  r2 = (r2-1.0e-14) ? r2 : 1.0e-14;
	  
          /* Propensity vector */
	  TIME = *T;
	  Propensity ( NmlcV, NmlcF, SCT, A );
	  
          /* Cumulative sum of propensities */
	  for (i=1; i<NREACT; i++)
            A[i] = A[i-1]+A[i];
          
	  /* Index of next reaction */
	  x = r1*A[NREACT-1];
	  for ( i = 0; i<NREACT; i++)
	    if (A[i] >= x) {
              m = i+1;
	      break;
	    }
	  
          /* Update T with time to next reaction */
          *T = *T - log(r2)/A[NREACT-1];

          /* Update state vector after reaction m */
	  MoleculeChange( m, NmlcV );
	  
     } /* for event */
    
} /* Gillespie */

File: WRF/chem/KPP/kpp/kpp-2.1/int/gillespie.f90
MODULE KPP_ROOT_Integrator

  USE KPP_ROOT_Parameters, ONLY : NVAR, NFIX, NREACT 
  USE KPP_ROOT_Global, ONLY : TIME, RCONST, Volume 
  USE KPP_ROOT_Stoichiom  
  USE KPP_ROOT_Stochastic
  USE KPP_ROOT_Rates
  IMPLICIT NONE

CONTAINS

!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  SUBROUTINE Gillespie(Nevents, T, SCT, NmlcV, NmlcF)
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!   Gillespie stochastic integration
!   INPUT:
!       Nevents = no. of individual reaction events to be simulated
!       SCT     = stochastic rate constants
!       T       = time
!       NmlcV, NmlcF = no. of molecules for variable and fixed species
!   OUTPUT:
!       T       = updated time (after Nevents reactions)
!       NmlcV   = updated no. of molecules for variable species
!      
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      IMPLICIT NONE
      KPP_REAL:: T
      INTEGER :: Nevents     
      INTEGER :: NmlcV(NVAR), NmlcF(NFIX)
      INTEGER :: i, m, issa
      REAL   :: r1, r2
      KPP_REAL :: A(NREACT), SCT(NREACT), x
   
      DO issa = 1, Nevents

          ! Uniformly distributed random numbers
          CALL RANDOM_NUMBER(r1)
          CALL RANDOM_NUMBER(r2)
          ! Avoid log of zero
          r2 = MAX(r2,1.e-14)
          
          ! Propensity vector
          CALL  Propensity ( NmlcV, NmlcF, SCT, A )
          ! Cumulative sum of propensities
          DO i = 2, NREACT
            A(i) = A(i-1)+A(i);
          END DO
          
          ! Index of next reaction
          x = r1*A(NREACT)
          DO i = 1, NREACT
            IF (A(i)>=x) THEN
              m = i;
              EXIT
            END IF
          END DO
          
          ! Update time with time to next reaction
          T = T - LOG(r2)/A(NREACT);

          ! Update state vector
          CALL MoleculeChange( m, NmlcV )
        
     END DO
    
  CONTAINS

    SUBROUTINE PropensityTemplate( T, NmlcV, NmlcF, Prop )
      KPP_REAL, INTENT(IN)  :: T
      INTEGER, INTENT(IN)   :: NmlcV(NVAR), NmlcF(NFIX)
      KPP_REAL, INTENT(OUT) :: Prop(NREACT)
      KPP_REAL :: Tsave, SCT(NREACT)
    ! Update the stochastic reaction rates, which may be time dependent 
      Tsave = TIME
      TIME = T
      CALL Update_RCONST()
      CALL StochasticRates( RCONST, Volume, SCT )   
      CALL Propensity ( NmlcV, NmlcF, SCT, Prop )
      TIME = Tsave
    END SUBROUTINE PropensityTemplate    
    
  END SUBROUTINE Gillespie

END MODULE KPP_ROOT_Integrator
